// @pakor/wacap-wrapper/dist/core/session-registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionRegistry = void 0;
const session_1 = require("./session");
/**
 * Manages lifecycle of Session instances and ensures isolation.
 */
class SessionRegistry {
    baseConfig;
    storageAdapter;
    globalBus;
    sessions = new Map();
    constructor(baseConfig, storageAdapter, globalBus) {
        this.baseConfig = baseConfig;
        this.storageAdapter = storageAdapter;
        this.globalBus = globalBus;
    }
    get(sessionId) {
        return this.sessions.get(sessionId);
    }
    has(sessionId) {
        return this.sessions.has(sessionId);
    }
    listIds() {
        return Array.from(this.sessions.keys());
    }
    all() {
        return this.sessions;
    }
    async create(sessionId, override) {
        if (this.sessions.has(sessionId)) {
            const existing = this.sessions.get(sessionId);
            if (existing.isActive()) {
                return existing;
            }
            // remove inactive session before recreating it
            this.sessions.delete(sessionId);
        }
        const merged = { ...this.baseConfig, ...override };
        const session = new session_1.Session(sessionId, merged, this.storageAdapter, this.globalBus);
        this.sessions.set(sessionId, session);
        await session.start();
        return session;
    }
    async destroy(sessionId) {
        const s = this.sessions.get(sessionId);
        if (!s)
            return;
        await s.stop();
        this.sessions.delete(sessionId);
    }
    /**
     * Logout and destroy a session (removes credentials)
     */
    async logout(sessionId) {
        const s = this.sessions.get(sessionId);
        if (!s)
            return;
        await s.logout();
        this.sessions.delete(sessionId);
    }
    async shutdownAll() {
        for (const [id, s] of this.sessions) {
            await s.stop();
        }
        this.sessions.clear();
    }
    async startByIds(sessionIds, override) {
        const started = [];
        for (const id of sessionIds) {
            const s = await this.create(id, override);
            started.push(s);
        }
        return started;
    }
    async restartAll() {
        const ids = this.listIds();
        await this.shutdownAll();
        return this.startByIds(ids);
    }
}
exports.SessionRegistry = SessionRegistry;
