// @pakor/wacap-wrapper/dist/handlers/message.handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerMessageHandlers = registerMessageHandlers;
const types_1 = require("../types");
function registerMessageHandlers(ctx) {
    const { socket, storageAdapter, sessionId, emit, touchActivity } = ctx;
    socket.ev.on('messages.upsert', async ({ messages }) => {
        for (const message of messages) {
            touchActivity();
            await storageAdapter.saveMessage(sessionId, message);
            const eventData = {
                message,
                isFromMe: message.key.fromMe,
                messageType: Object.keys(message.message || {})[0],
                body: getMessageBody(message),
                from: message.key.remoteJid,
            };
            if (message.key.fromMe) {
                emit(types_1.WacapEventType.MESSAGE_SENT, eventData);
            }
            else {
                emit(types_1.WacapEventType.MESSAGE_RECEIVED, eventData);
            }
        }
    });
    socket.ev.on('messages.update', async (updates) => {
        for (const update of updates) {
            touchActivity();
            emit(types_1.WacapEventType.MESSAGE_UPDATE, {
                message: update,
            });
        }
    });
    socket.ev.on('messages.delete', async () => {
        touchActivity();
        emit(types_1.WacapEventType.MESSAGE_DELETE, {});
    });
    socket.ev.on('chats.upsert', async (chats) => {
        for (const chat of chats) {
            await storageAdapter.saveChat(sessionId, chat);
        }
        touchActivity();
        emit(types_1.WacapEventType.CHAT_UPDATE, {});
    });
}
function getMessageBody(message) {
    const messageContent = message.message;
    if (!messageContent)
        return undefined;
    return (messageContent.conversation ||
        messageContent.extendedTextMessage?.text ||
        messageContent.imageMessage?.caption ||
        messageContent.videoMessage?.caption ||
        messageContent.documentMessage?.caption ||
        messageContent.buttonsResponseMessage?.selectedButtonId ||
        messageContent.listResponseMessage?.singleSelectReply?.selectedRowId ||
        messageContent.templateButtonReplyMessage?.selectedId ||
        messageContent.interactiveResponseMessage?.body?.text ||
        messageContent.editedMessage?.message?.protocolMessage?.editedMessage?.conversation ||
        messageContent.editedMessage?.message?.protocolMessage?.editedMessage?.extendedTextMessage?.text ||
        undefined);
}
