// @pakor/wacap-wrapper/dist/handlers/wrapper.helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatJid = formatJid;
exports.formatJids = formatJids;
exports.getSessionOrThrow = getSessionOrThrow;
exports.getSocketOrThrow = getSocketOrThrow;
exports.sendText = sendText;
exports.sendMedia = sendMedia;
exports.createGroup = createGroup;
exports.updateGroupParticipants = updateGroupParticipants;
exports.checkNumberStatus = checkNumberStatus;
exports.checkNumbersStatus = checkNumbersStatus;
exports.getProfilePicture = getProfilePicture;
exports.getContactInfo = getContactInfo;
exports.sendLocation = sendLocation;
exports.sendContact = sendContact;
exports.sendContacts = sendContacts;
exports.sendReaction = sendReaction;
exports.removeReaction = removeReaction;
exports.sendPoll = sendPoll;
exports.sendButtons = sendButtons;
exports.sendList = sendList;
exports.markAsRead = markAsRead;
exports.sendPresence = sendPresence;
exports.getGroupInfo = getGroupInfo;
exports.updateGroupSubject = updateGroupSubject;
exports.updateGroupDescription = updateGroupDescription;
exports.updateGroupSettings = updateGroupSettings;
exports.leaveGroup = leaveGroup;
exports.getGroupInviteCode = getGroupInviteCode;
exports.revokeGroupInviteCode = revokeGroupInviteCode;
exports.joinGroupViaCode = joinGroupViaCode;
exports.blockContact = blockContact;
exports.unblockContact = unblockContact;
exports.updateProfileStatus = updateProfileStatus;
exports.updateProfileName = updateProfileName;
exports.getChats = getChats;
exports.archiveChat = archiveChat;
exports.muteChat = muteChat;
exports.pinChat = pinChat;
exports.starMessage = starMessage;
exports.deleteMessage = deleteMessage;
exports.forwardMessage = forwardMessage;
exports.getBusinessProfile = getBusinessProfile;
/**
 * Format nomor telepon atau group ID ke JID WhatsApp
 *
 * @param input - Nomor telepon, group ID, atau JID lengkap
 * @param isGroup - Apakah ini group (default: auto-detect)
 * @returns JID yang sudah diformat
 *
 * @example
 * formatJid('08123456789') // '628123456789@s.whatsapp.net'
 * formatJid('628123456789') // '628123456789@s.whatsapp.net'
 * formatJid('+62 812-3456-789') // '628123456789@s.whatsapp.net'
 * formatJid('628123456789@s.whatsapp.net') // '628123456789@s.whatsapp.net' (unchanged)
 * formatJid('123456789', true) // '123456789@g.us'
 * formatJid('123456789@g.us') // '123456789@g.us' (unchanged)
 */
function formatJid(input, isGroup) {
    if (!input) {
        throw new Error('JID/nomor tidak boleh kosong');
    }
    // Trim whitespace
    let cleaned = input.trim();
    // Jika sudah format JID lengkap, return langsung
    // Support: @s.whatsapp.net, @g.us, @lid (Linked ID), @broadcast
    if (cleaned.endsWith('@s.whatsapp.net') ||
        cleaned.endsWith('@g.us') ||
        cleaned.endsWith('@lid') ||
        cleaned.endsWith('@broadcast')) {
        return cleaned;
    }
    // Auto-detect group jika mengandung '-' di tengah (format group ID WhatsApp)
    // Group ID biasanya format: 123456789-1234567890
    const looksLikeGroup = isGroup ?? /^\d+-\d+$/.test(cleaned);
    if (looksLikeGroup) {
        return `${cleaned}@g.us`;
    }
    // Format nomor telepon
    // Hapus semua karakter non-digit
    cleaned = cleaned.replace(/\D/g, '');
    // Handle format Indonesia
    if (cleaned.startsWith('0')) {
        // 08xxx -> 628xxx
        cleaned = '62' + cleaned.substring(1);
    }
    else if (cleaned.startsWith('8') && cleaned.length >= 9 && cleaned.length <= 13) {
        // 8xxx -> 628xxx (asumsi Indonesia jika 9-13 digit dimulai dengan 8)
        cleaned = '62' + cleaned;
    }
    return `${cleaned}@s.whatsapp.net`;
}
/**
 * Format array of JIDs
 */
function formatJids(inputs, isGroup) {
    return inputs.map(input => formatJid(input, isGroup));
}
/**
 * Resolve a session from registry or throw descriptive error.
 */
function getSessionOrThrow(registry, sessionId) {
    const session = registry.get(sessionId);
    if (!session) {
        throw new Error(`Session ${sessionId} not found`);
    }
    return session;
}
/**
 * Resolve socket from session or throw.
 */
function getSocketOrThrow(session) {
    const socket = session.getSocket();
    if (!socket) {
        throw new Error(`Session ${session.getInfo().sessionId} is not connected`);
    }
    return socket;
}
/**
 * Send text via registry
 * @param jid - Nomor telepon atau JID (akan di-format otomatis)
 */
async function sendText(registry, sessionId, jid, text, options) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    // Auto-format JID
    const formattedJid = formatJid(jid);
    const message = { text };
    if (options?.quoted) {
        return socket.sendMessage(formattedJid, message, { quoted: options.quoted });
    }
    if (options?.mentions) {
        // Format mentions juga
        const formattedMentions = formatJids(options.mentions);
        return socket.sendMessage(formattedJid, { text, mentions: formattedMentions });
    }
    return socket.sendMessage(formattedJid, message);
}
/**
 * Build media content block based on mimetype
 */
function buildMediaContent(media) {
    const content = {};
    const mimetype = media.mimetype || '';
    if (mimetype.startsWith('image/')) {
        content.image = media.url || media.buffer;
        if (media.caption)
            content.caption = media.caption;
    }
    else if (mimetype.startsWith('video/')) {
        content.video = media.url || media.buffer;
        if (media.caption)
            content.caption = media.caption;
    }
    else if (mimetype.startsWith('audio/')) {
        content.audio = media.url || media.buffer;
        content.mimetype = mimetype;
    }
    else {
        content.document = media.url || media.buffer;
        content.mimetype = mimetype;
        if (media.fileName)
            content.fileName = media.fileName;
        if (media.caption)
            content.caption = media.caption;
    }
    return content;
}
/**
 * Send media via registry
 * @param jid - Nomor telepon atau JID (akan di-format otomatis)
 */
async function sendMedia(registry, sessionId, jid, media) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    // Auto-format JID
    const formattedJid = formatJid(jid);
    const content = buildMediaContent(media);
    return socket.sendMessage(formattedJid, content);
}
/**
 * Create a new WhatsApp group
 * @param participantJids - Array nomor telepon atau JID (akan di-format otomatis)
 */
async function createGroup(registry, sessionId, subject, participantJids) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    // Auto-format participant JIDs
    const formattedParticipants = formatJids(participantJids);
    return socket.groupCreate(subject, formattedParticipants);
}
/**
 * Update participants of an existing group
 * @param groupJid - Group ID atau JID (akan di-format otomatis)
 * @param participantJids - Array nomor telepon atau JID (akan di-format otomatis)
 */
async function updateGroupParticipants(registry, sessionId, groupJid, participantJids, action) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    // Auto-format JIDs
    const formattedGroupJid = formatJid(groupJid, true);
    const formattedParticipants = formatJids(participantJids);
    return socket.groupParticipantsUpdate(formattedGroupJid, formattedParticipants, action);
}
// ============================================================
// FITUR TAMBAHAN (WAHA-like features)
// ============================================================
/**
 * Check if a phone number is registered on WhatsApp
 */
async function checkNumberStatus(registry, sessionId, phoneNumber) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(phoneNumber);
    const [result] = await socket.onWhatsApp(formattedJid.replace('@s.whatsapp.net', ''));
    return {
        exists: !!result?.exists,
        jid: result?.jid || formattedJid,
    };
}
/**
 * Check multiple numbers at once
 */
async function checkNumbersStatus(registry, sessionId, phoneNumbers) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const numbers = phoneNumbers.map((n) => formatJid(n).replace('@s.whatsapp.net', ''));
    const results = await socket.onWhatsApp(...numbers);
    return phoneNumbers.map((number, index) => ({
        number,
        exists: !!results[index]?.exists,
        jid: results[index]?.jid || formatJid(number),
    }));
}
/**
 * Get profile picture URL
 */
async function getProfilePicture(registry, sessionId, jid, highRes = true) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    try {
        const url = await socket.profilePictureUrl(formattedJid, highRes ? 'image' : 'preview');
        return url || null;
    }
    catch {
        return null;
    }
}
/**
 * Get contact info
 */
async function getContactInfo(registry, sessionId, jid) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    // Get status
    let status;
    try {
        const statusResult = await socket.fetchStatus(formattedJid);
        status = statusResult?.status;
    }
    catch {
        // Status not available
    }
    // Get profile picture
    let imgUrl;
    try {
        imgUrl = await socket.profilePictureUrl(formattedJid, 'image');
    }
    catch {
        // No profile picture
    }
    return {
        jid: formattedJid,
        status,
        imgUrl,
    };
}
/**
 * Send location message
 */
async function sendLocation(registry, sessionId, jid, latitude, longitude, options) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    return socket.sendMessage(formattedJid, {
        location: {
            degreesLatitude: latitude,
            degreesLongitude: longitude,
            name: options?.name,
            address: options?.address,
        },
    });
}
/**
 * Send contact/vCard
 */
async function sendContact(registry, sessionId, jid, contact) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    const vcard = `BEGIN:VCARD
VERSION:3.0
FN:${contact.name}
TEL;type=CELL;type=VOICE;waid=${contact.phone.replace(/\D/g, '')}:${contact.phone}
END:VCARD`;
    return socket.sendMessage(formattedJid, {
        contacts: {
            displayName: contact.name,
            contacts: [{ vcard }],
        },
    });
}
/**
 * Send multiple contacts
 */
async function sendContacts(registry, sessionId, jid, contacts) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    const vcards = contacts.map((c) => `BEGIN:VCARD
VERSION:3.0
FN:${c.name}
TEL;type=CELL;type=VOICE;waid=${c.phone.replace(/\D/g, '')}:${c.phone}
END:VCARD`);
    return socket.sendMessage(formattedJid, {
        contacts: {
            displayName: contacts.length === 1 ? contacts[0].name : `${contacts.length} contacts`,
            contacts: vcards.map((vcard) => ({ vcard })),
        },
    });
}
/**
 * Send reaction to a message
 */
async function sendReaction(registry, sessionId, jid, messageId, emoji) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    return socket.sendMessage(formattedJid, {
        react: {
            text: emoji,
            key: {
                remoteJid: formattedJid,
                id: messageId,
            },
        },
    });
}
/**
 * Remove reaction from a message
 */
async function removeReaction(registry, sessionId, jid, messageId) {
    return sendReaction(registry, sessionId, jid, messageId, '');
}
/**
 * Send poll
 */
async function sendPoll(registry, sessionId, jid, name, options, selectableCount = 1) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    return socket.sendMessage(formattedJid, {
        poll: {
            name,
            values: options,
            selectableCount,
        },
    });
}
/**
 * Send button message (deprecated in WhatsApp, may not work)
 */
async function sendButtons(registry, sessionId, jid, text, buttons, footer) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    return socket.sendMessage(formattedJid, {
        text,
        footer,
        buttons: buttons.map((b) => ({
            buttonId: b.id,
            buttonText: { displayText: b.text },
            type: 1,
        })),
    });
}
/**
 * Send list message
 */
async function sendList(registry, sessionId, jid, title, text, buttonText, sections, footer) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    return socket.sendMessage(formattedJid, {
        text,
        footer,
        title,
        buttonText,
        sections,
    });
}
/**
 * Mark message as read
 */
async function markAsRead(registry, sessionId, jid, messageIds) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    await socket.readMessages(messageIds.map((id) => ({
        remoteJid: formattedJid,
        id,
    })));
}
/**
 * Send presence update (typing, recording, online, offline)
 */
async function sendPresence(registry, sessionId, jid, presence) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = jid ? formatJid(jid) : undefined;
    await socket.sendPresenceUpdate(presence, formattedJid);
}
/**
 * Get group metadata
 */
async function getGroupInfo(registry, sessionId, groupId) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(groupId, true);
    const metadata = await socket.groupMetadata(formattedJid);
    return {
        id: metadata.id,
        subject: metadata.subject,
        description: metadata.desc,
        owner: metadata.owner,
        creation: metadata.creation,
        participants: metadata.participants.map((p) => ({
            id: p.id,
            admin: p.admin,
        })),
        announce: metadata.announce,
        restrict: metadata.restrict,
    };
}
/**
 * Update group subject (name)
 */
async function updateGroupSubject(registry, sessionId, groupId, subject) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(groupId, true);
    await socket.groupUpdateSubject(formattedJid, subject);
}
/**
 * Update group description
 */
async function updateGroupDescription(registry, sessionId, groupId, description) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(groupId, true);
    await socket.groupUpdateDescription(formattedJid, description);
}
/**
 * Update group settings
 */
async function updateGroupSettings(registry, sessionId, groupId, setting) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(groupId, true);
    await socket.groupSettingUpdate(formattedJid, setting);
}
/**
 * Leave a group
 */
async function leaveGroup(registry, sessionId, groupId) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(groupId, true);
    await socket.groupLeave(formattedJid);
}
/**
 * Get group invite code
 */
async function getGroupInviteCode(registry, sessionId, groupId) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(groupId, true);
    return socket.groupInviteCode(formattedJid);
}
/**
 * Revoke group invite code
 */
async function revokeGroupInviteCode(registry, sessionId, groupId) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(groupId, true);
    return socket.groupRevokeInvite(formattedJid);
}
/**
 * Join group via invite code
 */
async function joinGroupViaCode(registry, sessionId, inviteCode) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    // Remove URL prefix if present
    const code = inviteCode.replace('https://chat.whatsapp.com/', '');
    return socket.groupAcceptInvite(code);
}
/**
 * Block a contact
 */
async function blockContact(registry, sessionId, jid) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    await socket.updateBlockStatus(formattedJid, 'block');
}
/**
 * Unblock a contact
 */
async function unblockContact(registry, sessionId, jid) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    await socket.updateBlockStatus(formattedJid, 'unblock');
}
/**
 * Update profile status
 */
async function updateProfileStatus(registry, sessionId, status) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    await socket.updateProfileStatus(status);
}
/**
 * Update profile name
 */
async function updateProfileName(registry, sessionId, name) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    await socket.updateProfileName(name);
}
/**
 * Get all chats
 */
async function getChats(registry, sessionId) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    // Baileys stores chats in memory after sync
    const store = socket.store;
    if (store?.chats) {
        return Array.from(store.chats.all());
    }
    return [];
}
/**
 * Archive a chat
 */
async function archiveChat(registry, sessionId, jid, archive = true) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    await socket.chatModify({ archive }, formattedJid);
}
/**
 * Mute a chat
 */
async function muteChat(registry, sessionId, jid, muteEndTime) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    await socket.chatModify({ mute: muteEndTime }, formattedJid);
}
/**
 * Pin a chat
 */
async function pinChat(registry, sessionId, jid, pin = true) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    await socket.chatModify({ pin }, formattedJid);
}
/**
 * Star a message
 */
async function starMessage(registry, sessionId, jid, messageId, star = true) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    await socket.chatModify({
        star: {
            messages: [{ id: messageId, fromMe: false }],
            star,
        },
    }, formattedJid);
}
/**
 * Delete message for me
 */
async function deleteMessage(registry, sessionId, jid, messageId, forEveryone = false) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    if (forEveryone) {
        await socket.sendMessage(formattedJid, {
            delete: {
                remoteJid: formattedJid,
                id: messageId,
                participant: undefined,
            },
        });
    }
    else {
        await socket.chatModify({
            clear: { messages: [{ id: messageId, fromMe: false, timestamp: Date.now() }] },
        }, formattedJid);
    }
}
/**
 * Forward a message
 */
async function forwardMessage(registry, sessionId, jid, message, forceForward = false) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    return socket.sendMessage(formattedJid, {
        forward: message,
        force: forceForward,
    });
}
/**
 * Get business profile
 */
async function getBusinessProfile(registry, sessionId, jid) {
    const session = getSessionOrThrow(registry, sessionId);
    const socket = getSocketOrThrow(session);
    const formattedJid = formatJid(jid);
    try {
        return await socket.getBusinessProfile(formattedJid);
    }
    catch {
        return null;
    }
}
